### PR TITLE
Fix SSP restoring in edge cases

### PR DESCRIPTION
**Pull request by janvorli**
There are edge cases when the SSP restoring for continuation after a catch handler completes doesn't work correctly. The problem is caused by the fact that we scan for the Rip of the frame handling the exception on the shadow stack to find where to restore it, and in those edge cases, the same address can be there multiple times and the first occurence is not the right one. For example, when an exception is thrown from a catch handler, it escapes the handler and the handler for the escaped exception is in the same method as the one that invoked the handler.

This change fixes it by finding the SSP of the first managed frame where we search for the handler and then updating the SSP with every unwind. The SSP is stored in the REGDISPLAY. So when we reach the CallCatchFunclet, the REGDISPLAY contains the SSP to restore.

There was also one more issue with restoring the SSP. I turned out that the incsspq instruction uses only the lowest 8 bits of the argument to increment the SSP, so the ClrRestoreNonVolatileContextWorker needs to have a loop that repeats that instruction in case we need to move it by more than 255 slots.

### CODE REVIEW

File: src/coreclr/vm/amd64/cgenamd64.cpp [hunk 1]
```diff
@@ -108,13 +108,20 @@ void InlinedCallFrame::UpdateRegDisplay(const PREGDISPLAY pRD, bool updateFloats
     if (updateFloats)
     {
         UpdateFloatingPointRegisters(pRD);
+        // The float updating unwinds the stack so the pRD->pCurrentContext->Rip contains correct unwound Rip
+        // This is used for exception handling and the Rip extracted from m_pCallerReturnAddress is slightly
+        // off, which causes problem with searching for the return address on shadow stack on x64, so
+        // we keep the value from the unwind.
```
Comment (VSadov) [inline src/coreclr/vm/amd64/cgenamd64.cpp:114]: Was it a subtle bug that the Rip was off, or we just did not care about the correct IP in this case?

**Comment by jkotas** [comment]
Do we have test coverage for these corner cases?

**Comment by janvorli** [comment]
> Do we have test coverage for these corner cases?

There is no way to really test those, as Windows would happily run and silently fixup things.

File: src/coreclr/vm/amd64/cgenamd64.cpp (hunk 1 above)
Comment (janvorli) [reply src/coreclr/vm/amd64/cgenamd64.cpp:114]: That's actually expected due to the way how the InlinedCallFrame is created in some cases. Neither the GC stack walk nor the new EH cared, but it caused problem with the newly added search for the first managed frame, as it sometimes is one that called the pinvoke (QCALL).

**Comment by VSadov** [comment]
>> Do we have test coverage for these corner cases?

> There is no way to really test those, as Windows would happily run and silently fixup things.

Also testing the topmost SSP against what we can figure from SP is rarely helpful either. The top item is often correct, while we did not pop the things properly and the SSP keeps growing, for example. 

There is some sensitivity to wrong SSP in hijacking as OS unhijacks using SSP and we use our own stashed value, so incorrect SSP may lead to asserts. However observing this requires a fairly tight race with suspension. 
I wonder if running GC stress with CET enabled would trip on this more often. Although it would still be mostly about the top of the stack, which could be correct while the rest is not.

**Comment by VSadov** [comment]
Maybe there is a way to disable the OS fixup behavior, but then, I'd not be surprised something beyond our control requires it and nothing will run in such mode, even if it was possible to set up.

File: src/coreclr/vm/amd64/Context.S [hunk 2]
```diff
@@ -44,7 +44,14 @@ NESTED_ENTRY ClrRestoreNonvolatileContextWorker, _TEXT, NoHandler
         rdsspq  rax
         sub     r11, rax
         shr     r11, 3
-        incsspq r11
+        // the incsspq instruction uses only the lowest 8 bits of the argument, so we need to loop in case the increment is larger than 255
+        mov     rax, 255
+    Update_Loop:
+        cmp     r11, rax
+        cmovb   rax, r11
+        incsspq rax
```
Comment (VSadov) [inline src/coreclr/vm/amd64/Context.S:52]: This is indeed documented as only bits 0-7 are used, regardless of operand size.

Perhaps there is a way to save a byte of assembly by doing `INCSSPD eax`    :-)

Comment (janvorli) [reply src/coreclr/vm/amd64/Context.S:52]: The incsspd increments the SSP by 4, not by 8  :-)

**Comment by janvorli** [comment]
> Maybe there is a way to disable the OS fixup behavior

There is not, I've explicitly asked Windows folks about it.

File: src/coreclr/vm/amd64/Context.S (hunk 2 above)
Comment (VSadov) [reply src/coreclr/vm/amd64/Context.S:52]: I'd also make it use bits 0-3, because why not ...

**Review by VSadov — APPROVED** [review]
LGTM. Thanks!

**Comment by jkotas** [comment]
Is it possible for the shadow stack to overflow before this fix - can we use that to build regression test? Or can the effects of this fix be observed by better performance?

If there is really no way to observe the effects of this fix, I am wondering why it is needed. If it is the case, can we simplify things by leaving it to the OS to fix things up as necessary?

**Comment by VSadov** [comment]
> Is it possible for the shadow stack to overflow before this fix

yes, if we pop SSP to the fist expected IP match, but we need to pop to some other match, the overflow would be the ultimate result. After enough iterations. It could take many iterations though and very special crafted repro.

**Comment by jkotas** [comment]
We should create the special repro. I would expect that it should be a couple a viable outer loop test at least.

**Comment by VSadov** [comment]
It is possible that we already have tests when ssp grows, but we do not see that since the tests do not run long enough.  

I thought about some kind of asserting that shadow stack is roughly the same depth as the regular, and use in some random places, but not sure how that can be done in practice, considering native frames on stack.

**Comment by janvorli** [comment]
I will try to create a repro that would result in stack overflow without this change.

**Comment by janvorli** [comment]
@jkotas I have added a test that fails with (shadow) stack overflow before this fix and passes after.

File: src/tests/Regressions/coreclr/GitHub_104820/test104820.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            catch (Exception) when (depth == 100)
+            {
+            }
+        }
+    }
+
+    [Fact]
+    public static void TestEntryPoint()
+    {
+	ShadowStackPointerUpdateTest(100);
```
Comment (jkotas) [inline src/tests/Regressions/coreclr/GitHub_104820/test104820.cs:43]:
```suggestion
        ShadowStackPointerUpdateTest(100);
```
tabs vs. spaces

**Comment by jkotas** [comment]
> (shadow) stack overflow

Just curious - is there a special status code for shadow stack overflow or does it get reported as regular stack overflow?

**Comment by janvorli** [comment]
> Just curious - is there a special status code for shadow stack overflow or does it get reported as regular stack overflow?

It gets reported as regular stack overflow.

**Review by jkotas — APPROVED** [review]
Thank you for creating the test!

**Comment by VSadov** [comment]
BTW, The test also demonstrates that the fixup performed by Windows is not always correct. 

It is a good and perhaps the only possible backward-compat heuristic - pop to the nearest shadow stack item that matches the return site. However, the site may be recorded on the shadow stack multiple times and the code may be returning not to the lowermost occurrence.

**Comment by janvorli** [comment]
> BTW, The test also demonstrates that the fixup performed by Windows is not always correct.

I think that it will eventually fix itself after a couple of returns - one return might pick a wrong location, but then a next return will be wrong again, so yet another fixup will be made that finally corrects things.

**Pull request merged**